BitmovinConvivaAnalytics/Classes/ContentMetadataBuilder.swift

```diff
@@ -0,0 +1,191 @@
+//
+//  ContentMetadataBuilder.swift
+//  BitmovinConvivaAnalytics
+//
+//  Created by Bitmovin on 05.02.19.
+//  Copyright (c) 2019 Bitmovin. All rights reserved.
+//
+
+import Foundation
+import ConvivaSDK
+
+public struct MetadataOverrides {
+    // Can only be set once
+    public var assetName: String?
+
+    // Can only be set before playback started
+    public var viewerId: String?
+    public var streamType: StreamType?
+    public var applicationName: String?
+    public var custom: [String: Any]?
+    public var duration: Int?
+
+    // Dynamic
+    public var encodedFramerate: Int?
+    public var defaultResource: String?
+    public var streamUrl: String?
+
+    public init() {}
+}
+
+class ContentMetadataBuilder {
+    let logger: Logger
+    var contentMetadata: CISContentMetadata
+
+    // internal metadata fields to enable merging / overriding
+    var metadataOverrides: MetadataOverrides = MetadataOverrides()
+    var metadata: MetadataOverrides = MetadataOverrides()
+    var playbackStarted: Bool = false
+
+    init(logger: Logger) {
+        self.logger = logger
+        contentMetadata = CISContentMetadata()
+    }
+
+    public func setOverrides(_ metadataOverrides: MetadataOverrides) {
+        if playbackStarted {
+            logger.debugLog(
+                message: "[ Conviva Analytics ] Playback has started. Only some metadata attributes will be updated"
+            )
+        }
+
+        self.metadataOverrides = metadataOverrides
+    }
+
+    public func setPlaybackStarted(_ playbackStarted: Bool) {
+        self.playbackStarted = playbackStarted
+    }
+
+    public func build() -> CISContentMetadata {
+        if !playbackStarted {
+            // Asset name is only allowed to be set once
+            if contentMetadata.assetName == nil {
+                contentMetadata.assetName = assetName
+            }
+
+            contentMetadata.viewerId = viewerId
+            contentMetadata.applicationName = applicationName
+
+            if let type = streamType {
+                contentMetadata.streamType = type
+            }
+            if let dur = duration {
```

Comment:
would it be possible to use `self` here as well? I know we already talked about these namings...

Reply:
Should be possible